metrics/sinks/gcm/gcm.go

```diff
@@ -321,21 +324,39 @@ func CreateGCMSink(uri *url.URL) (core.DataSink, error) {
 		return nil, fmt.Errorf("invalid metrics parameter: %s", metrics)
 	}
 
-	if err := gce_util.EnsureOnGCE(); err != nil {
-		return nil, err
-	}
-
-	// Detect project ID
-	projectId, err := gce.ProjectID()
-	if err != nil {
-		return nil, err
-	}
+	var projectId string
+	var gcmService *gcm.Service
 
-	// Create Google Cloud Monitoring service.
-	client := oauth2.NewClient(oauth2.NoContext, google.ComputeTokenSource(""))
-	gcmService, err := gcm.New(client)
-	if err != nil {
-		return nil, err
+	if os.Getenv(gcpCredentialEnv) != "" {
+		// GCP credentials supplied, so assume we're not in GCE
+		if os.Getenv(gcpProjectEnv) == "" {
+			return nil, fmt.Errorf("no GCP project ID supplied")
+		} else {
+			projectId = os.Getenv(gcpProjectEnv)
+		}
+		glog.Infof("Google project ID is: %s", projectId)
+		// Create Google Cloud Monitoring service
+		if client, err := google.DefaultClient(oauth2.NoContext, gcm.MonitoringScope); err != nil {
+			return nil, err
+		} else {
+			gcmService, err = gcm.New(client)
+		}
+	} else {
+		// Ensure we're on GCE
+		if err := gce_util.EnsureOnGCE(); err != nil {
+			return nil, err
+		}
+		// Detect project ID
+		projectId, err = gce.ProjectID()
+		if err != nil {
+			return nil, err
+		}
+		// Create Google Cloud Monitoring service.
+		client := oauth2.NewClient(oauth2.NoContext, google.ComputeTokenSource(""))
```

Comment:
This is actually already done by `google.DefaultClient` when the credentials file isn't present. Looks like a bunch of that code can be reused.